### PR TITLE
MULTIARCH-5161: Power VS: Create VPEs as needed for Disconnected install

### PR DESCRIPTION
**Pull request by miyamotoh**
For Disconnected install on PowerVS, Virtual Private Endpoint Gateways (VPEs) must be created in the VPC that would be part of the infrastructure to allow traffic to the following private IBM Cloud endpoints;
- IAM
- COS (in the region)
- DNS services

Proposed code would first check if the necessary VPEs are already in place, comparing the target's CRN, and create one only when missing. 
Destroy logic for the below resources need to be developed (later);
- Reserved IPs (on the subnet)
- VPEs (in the VPC)

### CODE REVIEW

File: pkg/infrastructure/powervs/clusterapi/powervs.go [hunk 1]
```diff
@@ -163,6 +163,34 @@ func (p Provider) InfraReady(ctx context.Context, in clusterapi.InfraReadyInput)
...
+		if err != nil {
+			return fmt.Errorf("failed to obtain PowerVS client: %w", err)
+		}
+
+		name := fmt.Sprintf("%s-vpe-iam", in.InfraID)
+		vpcID := *powerVSCluster.Status.VPC.ID
+		logrus.Debugf("InfraReady: Ensuring VPE gateway for IAM in VPC %v", vpcID)
+		rgID := *powerVSCluster.Status.ResourceGroup.ID
+		subnetID := *powerVSCluster.Status.VPCSubnet[*powerVSCluster.Spec.VPCSubnets[1].Name].ID
+		_, err = client.CreateVirtualPrivateEndpointGateway(ctx, name, vpcID, subnetID, rgID, "crn:v1:bluemix:public:iam-svcs:global:::endpoint:private.iam.cloud.ibm.com")
```
Comment (mjturek) [inline pkg/infrastructure/powervs/clusterapi/powervs.go:180]: This isn't needed for the private case, it's needed for the disconnected case. So there should be an inner check here that ImageContentDigests (or sources) is set.

Comment (mjturek) [reply pkg/infrastructure/powervs/clusterapi/powervs.go:180]: Reading closer this seems to be just a test, so you probably are going to make this cleaner once you start considering other VPEs that are required for disconnected :)

**Comment by mjturek** [comment]
We also need DNSServices for disconnected (the ingress operator uses it)

File: pkg/infrastructure/powervs/clusterapi/powervs.go (hunk 1 above)
Comment (miyamotoh) [reply pkg/infrastructure/powervs/clusterapi/powervs.go:180]: Line 168 in the proposed change is what I _think_ signifies the Disconnected install, on top of line 167 checking it's an `Internal` (=private) install. So, this is real code that'd create IAM VPE for Disconnected install, though it'd also execute, if you specify `imageContentSource` pointing to public `quay.io` repos as your "mirror," and it'd perhaps fail. So, that's the early discussion point I'm looking to have. 🙂

Comment (mjturek) [reply pkg/infrastructure/powervs/clusterapi/powervs.go:180]: ah sorry i see it now

File: pkg/asset/installconfig/powervs/metadata.go [hunk 2]
```diff
@@ -491,3 +492,31 @@ func leftInContext(ctx context.Context) time.Duration {
...
+	}
+
+	name = fmt.Sprintf("%s-vpe-cos", infraID)
+	s3Crn := fmt.Sprintf("crn:v1:bluemix:public:cloud-object-storage:global:::endpoint:s3.direct.%s.cloud-object-storage.appdomain.cloud", region)
+	logrus.Debugf("InfraReady: Ensuring VPE gateway for COS %v", s3Crn)
+	_, err = client.CreateVirtualPrivateEndpointGateway(ctx, name, vpcID, subnetID, groupID, s3Crn)
+	if err != nil {
+		return fmt.Errorf("failed to create VPE: %w", err)
+	}
+
```
Comment (mjturek) [inline pkg/asset/installconfig/powervs/metadata.go:520]: Please add a VPE for DNSServices

File: pkg/infrastructure/powervs/clusterapi/powervs.go [hunk 3]
```diff
@@ -163,6 +163,19 @@ func (p Provider) InfraReady(ctx context.Context, in clusterapi.InfraReadyInput)
 	if err != nil {
 		return fmt.Errorf("failed to add ping security group rule: %w", err)
 	}
+
+	if in.InstallConfig.Config.Publish == types.InternalPublishingStrategy &&
+		(len(in.InstallConfig.Config.ImageDigestSources) > 0 || len(in.InstallConfig.Config.DeprecatedImageContentSources) > 0) {
+		vpcID := *powerVSCluster.Status.VPC.ID
+		logrus.Debugf("InfraReady: Ensuring VPE gateway for IAM in VPC %v", vpcID)
```
Comment (mjturek) [inline pkg/infrastructure/powervs/clusterapi/powervs.go:170]: This log is inaccurate, make it clear that multiple gateways are being created (IAM, COS, DNS)

File: pkg/asset/installconfig/powervs/metadata.go (hunk 2 above)
Comment (miyamotoh) [reply pkg/asset/installconfig/powervs/metadata.go:531]: Done. PTAL at the latest force-push.

File: pkg/infrastructure/powervs/clusterapi/powervs.go (hunk 3 above)
Comment (miyamotoh) [reply pkg/infrastructure/powervs/clusterapi/powervs.go]: Corrected in the latest force-push. PTAL.

File: pkg/asset/installconfig/powervs/metadata.go [hunk 4]
```diff
@@ -491,3 +493,105 @@ func leftInContext(ctx context.Context) time.Duration {
...
+		}
+	}
+
+	endpoint, err = m.fetchEndpointOverride(endpointOverrides, string(configv1.IBMCloudServiceResourceController))
+	if err != nil {
+		return fmt.Errorf("failed to fetch endpoint override found for RC: %w", err)
+	}
+	if endpoint == "" {
+		name := fmt.Sprintf("%s-vpe-rc", infraID)
+		rcCrn := "crn:v1:bluemix:public:resource-controller:global:::endpoint:private.resource-controller.cloud.ibm.com"
```
Comment (sadasu) [inline pkg/asset/installconfig/powervs/metadata.go:579]: Nit: Please add this as a constant.

File: pkg/asset/installconfig/powervs/metadata.go [hunk 5]
```diff
@@ -491,3 +493,105 @@ func leftInContext(ctx context.Context) time.Duration {
...
+		return err
+	}
+
+	endpoint, err := m.fetchEndpointOverride(endpointOverrides, "IAM")
+	if err != nil {
+		return fmt.Errorf("failed to fetch endpoint override found for IAM: %w", err)
+	}
+	if endpoint == "" {
+		name := fmt.Sprintf("%s-vpe-iam", infraID)
+		iamCrn := "crn:v1:bluemix:public:iam-svcs:global:::endpoint:private.iam.cloud.ibm.com"
```
Comment (sadasu) [inline pkg/asset/installconfig/powervs/metadata.go:537]: Nit: Please add this as a constant.

File: pkg/asset/installconfig/powervs/metadata.go [hunk 6]
```diff
@@ -491,3 +493,105 @@ func leftInContext(ctx context.Context) time.Duration {
...
+		}
+	}
+
+	endpoint, err = m.fetchEndpointOverride(endpointOverrides, "COS")
+	if err != nil {
+		return fmt.Errorf("failed to fetch endpoint override found for COS: %w", err)
+	}
+	if endpoint == "" {
+		name := fmt.Sprintf("%s-vpe-cos", infraID)
+		cosCrn := fmt.Sprintf("crn:v1:bluemix:public:cloud-object-storage:global:::endpoint:s3.direct.%s.cloud-object-storage.appdomain.cloud", region)
```
Comment (sadasu) [inline pkg/asset/installconfig/powervs/metadata.go:551]: Ditto

File: pkg/infrastructure/powervs/clusterapi/powervs.go [hunk 7]
```diff
@@ -163,6 +163,19 @@ func (p Provider) InfraReady(ctx context.Context, in clusterapi.InfraReadyInput)
 	if err != nil {
 		return fmt.Errorf("failed to add ping security group rule: %w", err)
 	}
+
+	if in.InstallConfig.Config.Publish == types.InternalPublishingStrategy &&
+		(len(in.InstallConfig.Config.ImageDigestSources) > 0 || len(in.InstallConfig.Config.DeprecatedImageContentSources) > 0) {
```
Comment (sadasu) [inline pkg/infrastructure/powervs/clusterapi/powervs.go:168]: Assuming we are within the deprecation window of `InstallConfig.Config.DeprecatedImageContentSources` would it be useful to set InstallConfig.Config.ImageDigestSources to InstallConfig.Config.DeprecatedImageContentSources when only InstallConfig.Config.DeprecatedImageContentSources is provided while validating InstallConfig? That way we don't have to check both values every time. 
When you actually stop support for InstallConfig.Config.DeprecatedImageContentSources, you wouldn't have to make changes in so many places.

File: pkg/asset/installconfig/powervs/metadata.go (hunk 6 above)
Comment (miyamotoh) [reply pkg/asset/installconfig/powervs/metadata.go]: Addressed in the latest force-push. PTAL.

File: pkg/asset/installconfig/powervs/metadata.go (hunk 5 above)
Comment (miyamotoh) [reply pkg/asset/installconfig/powervs/metadata.go]: Addressed in the latest force-push. PTAL.

File: pkg/asset/installconfig/powervs/metadata.go (hunk 4 above)
Comment (miyamotoh) [reply pkg/asset/installconfig/powervs/metadata.go]: Addressed in the latest force-push. PTAL.

**Comment by mjturek** [comment]
/retitle MULTIARCH-5161: Power VS: Create VPEs as needed for Disconnected install

**Review by mjturek — COMMENTED** [review]
/lgtm

File: pkg/asset/cluster/metadata.go [hunk 8]
```diff
@@ -97,7 +97,10 @@ func (m *Metadata) Generate(_ context.Context, parents asset.Parents) (err error
 	case vspheretypes.Name:
 		metadata.ClusterPlatformMetadata.VSphere = vsphere.Metadata(installConfig.Config)
 	case powervstypes.Name:
-		metadata.ClusterPlatformMetadata.PowerVS = powervs.Metadata(installConfig.Config, installConfig.PowerVS)
+		metadata.ClusterPlatformMetadata.PowerVS, err = powervs.Metadata(installConfig.Config, installConfig.PowerVS)
+		if err != nil {
+			return errors.Errorf("failed to initialize PowerVS: %v", err)
```
Comment (r4f4) [inline pkg/asset/cluster/metadata.go:102]: github.com/pkg/errors is deprecated:
```suggestion
			return fmt.Errorf("failed to initialize PowerVS: %w", err)
```

File: pkg/asset/installconfig/powervs/session.go [hunk 9]
```diff
@@ -427,14 +426,22 @@ func getEnv(envs []string) string {
...
+		"ResourceController": "", // FIXME CAPI recognizes "rc," but crashes if passed in...
+		"ResourceManager":    "rm",
+		"VPC":                "vpc",
+	}
 	overrides := make([]string, 0, len(cfg.ServiceEndpoints))
 	// CAPI expects name=url pairs of service endpoints
 	for _, endpoint := range cfg.ServiceEndpoints {
-		if capiSupported.Has(endpoint.Name) {
-			overrides = append(overrides, fmt.Sprintf("%s=%s", endpoint.Name, endpoint.URL))
+		if capiSupported[endpoint.Name] != "" {
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/session.go:443]: It's probably a good idea to do 
```suggestion
		if capiName, ok := capiSupported[endpoint.Name]; ok && name != "" {
```

File: pkg/asset/manifests/infrastructure.go [hunk 10]
```diff
@@ -311,6 +311,19 @@ func (i *Infrastructure) Generate(ctx context.Context, dependencies asset.Parent
 				URL:  service.URL,
 			})
 		}
+		if installConfig.Config.Publish == types.InternalPublishingStrategy &&
+			(len(installConfig.Config.ImageDigestSources) > 0 || len(installConfig.Config.DeprecatedImageContentSources) > 0) {
+			piRegion := installConfig.Config.PowerVS.Region
+			vpcRegion, err := powervs.VPCRegionForPowerVSRegion(piRegion)
+			if err != nil {
+				return errors.New("failed to determine VPC region")
```
Comment (r4f4) [inline pkg/asset/manifests/infrastructure.go:319]: Are you not bubbling up the error on purpose?
```suggestion
				return fmt.Errorf("failed to determine VPC region: %w", err)
```

File: pkg/asset/manifests/infrastructure.go [hunk 11]
```diff
@@ -311,6 +311,19 @@ func (i *Infrastructure) Generate(ctx context.Context, dependencies asset.Parent
...
+			}
+			cosRegion, err := powervs.COSRegionForPowerVSRegion(piRegion)
+			if err != nil {
+				return errors.New("failed to determine COS region")
```
Comment (r4f4) [inline pkg/asset/manifests/infrastructure.go:323]:
```suggestion
				return fmt.Errorf("failed to determine COS region: %w", err)
```

File: pkg/asset/manifests/infrastructure.go (hunk 10 above)
Comment (miyamotoh) [reply pkg/asset/manifests/infrastructure.go:319]: Not on purpose. Overside. Fixed.

File: pkg/asset/manifests/infrastructure.go (hunk 11 above)
Comment (miyamotoh) [reply pkg/asset/manifests/infrastructure.go:323]: Ditto.

**Review by r4f4 — COMMENTED** [review]
/approve

**Comment by sadasu** [comment]
/label acknowledge-critical-fixes-only

@mjturek can add the final label after review

**Pull request merged**